doc: add documentation

antonio-pedro99 opened the pull request:
This PR proposes the documentation for the MQTT bridge. It contains the following informations:
> How to run the Bridge
> How to use the Bridge
> How to configure the bridge.

Signed-off-by: Antonio Pedro <tonio.pedro99@gmail.com>

ppatierno commented on README.md:
```diff
@@ -4,3 +4,116 @@
+## MQTT bridge documentation
+
+### 1. MQTT Bridge Overview
+You can use the MQTT bridge to take advantage of the benefits of both MQTT and Kafka in data processing and IoT system scenarios.
```
Add one blank like at the top

ppatierno replied:
Even if tbh I don't see how much this sentence is useful here, I would just remove it.

ppatierno reviewed (commented):
I had a very first quick pass

kyguy commented on README.md:
````diff
@@ -4,3 +4,116 @@
+Once your configuration is ready, start the bridge using:
+
+```bash
+<application> --config-file config/application.properties --mapping-rules=path/to/mapping-rules
````
There is a placeholder  name, <application>,  because we don't know what the final application name will be?

ppatierno replied:
Yeah this line doesn't have much sense right now. We should package the application and provide a jar which is runnable, so that we can replace `<application>` with something meanigful.

antonio-pedro99 replied:
@ppatierno should we package the application now?

ppatierno replied:
Right now we don't have an application which does what it's aim for ... so would not make sense to package. 
At the same time, this line doesn't make sense.
Right now, maybe, we could avoid the section about running the bridge and adding it later when it will be possible to run. @kyguy what's your taken on it?

antonio-pedro99 replied:
for me sounds good removing it.

kyguy commented on README.md:
```diff
@@ -4,3 +4,116 @@
+
+#### 1.2 How it works
+
+The MQTT Bridge is a netty-based MQTT server that is responsible for handling all the MQTT client connections and connections management and is responsible for all MQTT-based operations like QoS, error handling, and other capabilities Netty offers.
```
```suggestion
The MQTT Bridge is a netty-based MQTT server that is responsible for handling all the MQTT client connections, connections management, and all MQTT-based operations like QoS, error handling, and other capabilities which Netty offers.
```

kyguy commented on README.md:
```diff
@@ -4,3 +4,116 @@
+
+The MQTT Bridge is a netty-based MQTT server that is responsible for handling all the MQTT client connections and connections management and is responsible for all MQTT-based operations like QoS, error handling, and other capabilities Netty offers.
+
+The MQTT Bridge provides two interfaces to integrate MQTT and Apache Kafka. The mapper maps incoming MQTT topics to Kafka topics. Once this mapping is completed, the MQTT Bridge will transform the incoming MQTT payload to a valid Kafka record and then send to the appropriate Kafka topic through the bridge kafka Producer.
```
Just a suggestion, feel free to ignore
```suggestion
The MQTT Bridge provides two interfaces to integrate MQTT and Apache Kafka. 
1. MQTT mapper: maps incoming MQTT topics to Kafka topics. 
2. MQTT Bridge:  transforms the incoming MQTT payload to a valid Kafka record and then sends them to the appropriate Kafka topic using a Kafka Producer.

```

kyguy commented on README.md:
```diff
@@ -4,3 +4,116 @@
+The MQTT Bridge provides two interfaces to integrate MQTT and Apache Kafka. The mapper maps incoming MQTT topics to Kafka topics. Once this mapping is completed, the MQTT Bridge will transform the incoming MQTT payload to a valid Kafka record and then send to the appropriate Kafka topic through the bridge kafka Producer.
+
+#### 1.2.1 The MqttKafkaMapper
+This interface is responsible for mapping incoming MQTT topic to Kafka topic. How this mapping is should be made must be defined in the `Topic Mapping Rules(TOMAR)`.
```
```suggestion
This interface is responsible for mapping incoming MQTT topic names to Kafka topic names. How this mapping is done must must be defined in the `Topic Mapping Rules(TOMAR)`.
```

kyguy commented on README.md:
```diff
@@ -4,3 +4,116 @@
+
+#### 1.2.2 Topic Mapping Rules(TOMAR)
+
+The TOMAR is a set of patterns the user provides, to which the MQTT Bridge will decide which Kafka topic to produce a message. 
```
```suggestion
The TOMAR is a set of patterns the user provides defining how the MQTT Bridge maps MQTT topic names to Kafka topic names.
```

kyguy commented on README.md:
```diff
@@ -4,3 +4,116 @@
+The MQTT bridge can be configured using the appropriate prefix. Example:
+- `bridge.` is the prefix used for general configuration of the Bridge.
+- `mqtt.` is the prefix used for MQTT configuration of the Bridge.
+- `kafka.` is the prefix used for Kafka configuration of the Bridge.
```
SHould we mention that they are MQTT server configuration and Kafka producer configurations?

antonio-pedro99 replied:
@kyguy sorry, I did not understand. I think it is already mentioning, `mqtt.` is the prefix used for `MQTT configuration` of the Bridge.

antonio-pedro99 replied:
@kyguy any insights for this?

kyguy replied:
Sorry I meant something like this
```suggestion
- `kafka.` is the prefix used for Kafka producer configurations of the Bridge.
```
if you think it is useful, if not, please ignore

kyguy replied:
I guess it is already implied

antonio-pedro99 replied:
Well, I think we are using this prefix not only for Kafka producer configuration.

ppatierno replied:
We actually have just the producer in our MQTT bridge. What other kind of Kafka configuration are you referring to? If you are referring to something line `kafka.bootstrap.servers` this is anyway a configuration for the producer.
We don't have it under `kafka.producer.` just because you got the code from the HTTP bridge where we decided to have common config across consumer and producer in a `kafka.` parameters.
In this case we could decide to not having a Kafka common section and having `kafka.producer.bootstrap.servers` but tbh I would leave as it is. Who knows if far in the future we'll decide to add an MQTT consumer part as well?

antonio-pedro99 commented on README.md:
```diff
@@ -4,3 +4,116 @@
+- `mqtt.` is the prefix used for MQTT configuration of the Bridge.
+- `kafka.` is the prefix used for Kafka configuration of the Bridge.
+### 2.1 Procedure to configure the MQTT Bridge
+1. Find the `application.properties` file in the `config` directory.
```
before line 91? Or before line 89?

ppatierno replied:
You should add a blank line after each title, after 91 ... this is valid for all the similar comments I made.

ppatierno replied:
Also, the user could put the application.properties file in any directory (this is why we have the --config-file option) so they don't need to "find" the file in the config directory, they can create their own.

antonio-pedro99 replied:
@ppatierno finding the file was just for reference.

antonio-pedro99 replied:
Assuming the user will clone the repo, he just needs to find it. But I will replace with more general thing.

ppatierno commented on README.md:
```diff
@@ -4,3 +4,119 @@
+
+The MQTT Bridge provides two interfaces to integrate MQTT and Apache Kafka. 
+1. MQTT mapper: maps incoming MQTT topics to Kafka topics. 
+2. MQTT Bridge:  transforms the incoming MQTT payload to a valid Kafka record and then sends them to the appropriate Kafka topic using a Kafka Producer.
```
for me the MQTT bridge is the entire application, it's not what you are explaining here. It's the entire bridge, including the MQTT server, the topic mapper, the Kafka producer and so on.

ppatierno commented on README.md:
```diff
@@ -4,3 +4,119 @@
+
+
+#### 1.2.1 The MqttKafkaMapper
+This interface is responsible for mapping incoming MQTT topic names to Kafka topic names. How this mapping is done must must be defined in the `Topic Mapping Rules(TOMAR)`.
```
Not sure a paragraph with just one line makes any sense, I would just talk about topic mapping rules and not about `MqttKafkaMapper` interface. It doesn't matter for users, they want to know how to map topics.

ppatierno commented on README.md:
```diff
@@ -4,3 +4,119 @@
+#### 1.2.1 The MqttKafkaMapper
+This interface is responsible for mapping incoming MQTT topic names to Kafka topic names. How this mapping is done must must be defined in the `Topic Mapping Rules(TOMAR)`.
+
+#### 1.2.2 Topic Mapping Rules(TOMAR)
```
Can you remember me why we need a TOMAR acronym here? From where it's coming?

ppatierno commented on README.md:
```diff
@@ -4,3 +4,119 @@
+#### 1.2.2 Topic Mapping Rules(TOMAR)
+
+The TOMAR is a set of patterns the user provides defining how the MQTT Bridge maps MQTT topic names to Kafka topic names.
+A Mapping Rule is a model that contains an `MQTT topic pattern` and a `Kafka topic template`. All the incoming MQTT message's topic should match an `MQTT topic pattern` in the `TOMAR` so that the bridge knows in which Kafka topic to produce this message. This Kafka topic is defined by a template, `kafka Topic template`. However, if the incoming MQTT message's topic does not match any pattern in t ...
```
Split paragraph with one sentence per line. This is important because help to simplify to review process, where reviewers can add feedback and comment on each sentence.

antonio-pedro99 replied:
@ppatierno  I think I got it wrong.

ppatierno commented on README.md:
````diff
@@ -4,3 +4,119 @@
+  }
+]
+```
+Let's go through each rule:
````
Before going through the example you should summarize the role of `+`, `#` and placeholders (what they are).

antonio-pedro99 replied:
@ppatierno yes, I am explaining about them while explaining the rules.

ppatierno replied:
I think they should be explained and summarized before. They could be lost while explaining an example. User should be able to come here and using the summary as a reference and not going through the examples to understand them.

antonio-pedro99 replied:
Sounds good!

ppatierno commented on README.md:
```diff
@@ -4,3 +4,119 @@
+- `bridge.` is the prefix used for general configuration of the Bridge.
+- `mqtt.` is the prefix used for MQTT configuration of the Bridge.
+- `kafka.` is the prefix used for Kafka configuration of the Bridge.
+### 2.1 Procedure to configure the MQTT Bridge
```
Tbh I am not sure we need a procedure here. You explained how to configure the bridge via the application.properties file, it's clear you have to open, edit and save :-)

antonio-pedro99 replied:
@ppatierno it is just to explain the possible configuration prefix. I think it is important to mention it.

ppatierno replied:
Just explain them, not going through a procedure where you are just saying to open, edit and save a file.

ppatierno commented on README.md:
```diff
@@ -4,3 +4,119 @@
+The following table describes the configuration properties that can be used to configure the MQTT Bridge.
+
+| Setting                        | Description                                     | E.g. Value              |
+| ------------------------------ | ----------------------------------------------- |-------------------------|
```
I don't think we want an e.g. Value column, but a "Default" column only.

antonio-pedro99 replied:
@ppatierno I agree.

ppatierno reviewed (changes requested):
A common comment to all of them is to use one sentence per line, it makes life easier to reviewers to make comments on each line and not to a whole paragraph.
Please rewrite this way while addressing my comments.

antonio-pedro99 commented on README.md:
```diff
@@ -4,3 +4,116 @@
+
+### Locally
+
+Clone the repository and update your dependencies using maven.
```
Sync the dependencies.

ppatierno replied:
I still don't get it, sorry. If you are referencing to download the dependency, it should be part of a section talking about "how to build" and running the corresponding maven command. It will be the one to download the dependency. It's not something the user has to do manually.

antonio-pedro99 replied:
this is out for now

ppatierno commented on README.md:
```diff
@@ -3,4 +3,135 @@
 
-This project provides a software component which acts as a bridge between [MQTT 3.1.1](http://docs.oasis-open.org/mqtt/mqtt/v3.1.1/os/mqtt-v3.1.1-os.html) and an [Apache Kafka®](https://kafka.apache.org/) cluster.
+This project provides a software component which acts as a bridge between [MQTT 3.1.1](http://docs.oasis-open.org/mqtt/mqtt/v3.1.1/os/mqtt-v3.1.1-os.html) 
+and an [Apache Kafka®](https://kafka.apache.org/) cluster.
```
why breaking in two lines here?

ppatierno commented on README.md:
```diff
@@ -3,4 +3,135 @@
+#### 1.1 How it works
+
+To enable a seamless integration between MQTT and Kafka, the MQTT Bridge provides a way to map MQTT topics to Kafka topics. 
+This mapping is done using a set of predefined patterns and kafka topic templates. The MQTT Bridge uses these patterns to map MQTT topics to Kafka topics. 
```
two sentences, two lines.

ppatierno commented on README.md:
```diff
@@ -3,4 +3,135 @@
+The TOMAR is a set of patterns the user provides defining how the MQTT Bridge maps MQTT topic names to Kafka topic names.
+A Mapping Rule is a model that contains an `MQTT topic pattern` and a `Kafka topic template`. 
+All the incoming MQTT message's topic should match an `MQTT topic pattern` in the `TOMAR` so that the bridge knows in which 
+Kafka topic to produce this message. This Kafka topic is defined by a template, `kafka Topic template`. 
```
the first sentence is broken in two lines (it shouldn't). The second sentence should be on its own.

ppatierno replied:
```suggestion
This Kafka topic is defined by a template, `kafka Topic template`. 
```

ppatierno commented on README.md:
```diff
@@ -3,4 +3,135 @@
+
+    This rule maps MQTT topics of the form "building/{building}/room/{room}/#" to the Kafka topic "building_{building}room{room}". 
+    The wildcard "#" represents one or more levels in the topic hierarchy. For example, if the MQTT topic is "building/A/room/1/floor/2", 
+    it will be mapped to the Kafka topic "building_A_room_1_floor_2".
```
one sentence, one line

ppatierno commented on README.md:
```diff
@@ -3,4 +3,135 @@
+
+    This rule maps any MQTT topic starting with "sensors/" followed by any number of levels in the hierarchy to the Kafka topic "sensor_others". 
+    The "#" wildcard matches any number of levels. For example, if the MQTT topic is "sensors/temperature/living-room", 
+    it will be mapped to the Kafka topic "sensor_others".
```
one sentence, one line

ppatierno commented on README.md:
```diff
@@ -3,4 +3,135 @@
+
+    This rule maps MQTT topics of the form "sensors/+/data" to the Kafka topic "sensor_data". The "+" 
+    wildcard matches a single level in the topic hierarchy. For example, if the MQTT topic is "sensors/temperature/data", 
+    it will be mapped to the Kafka topic "sensor_data".
```
one sentence, one line

ppatierno commented on README.md:
```diff
@@ -3,4 +3,135 @@
+
+These mapping rules demonstrate how different MQTT topics can be mapped to corresponding Kafka topics based on predefined patterns.
+The MQTT Bridge uses these rules to map MQTT topics to Kafka topics.
+The order in which the rules are defined is important. The MQTT Bridge will use the first rule that matches the MQTT topic. 
```
two sentence, two lines.

ppatierno commented on README.md:
```diff
@@ -3,4 +3,135 @@
+The order in which the rules are defined is important. The MQTT Bridge will use the first rule that matches the MQTT topic. 
+For example, if the MQTT topic is "sensors/temperature/data", 
+it will be mapped to the Kafka topic "sensor_data" because the third rule matches the MQTT topic. If the third rule was not defined,
+the MQTT Bridge would use the second rule to map the MQTT topic to the Kafka topic "sensor_others".
```
all on one line

ppatierno commented on README.md:
```diff
@@ -3,4 +3,135 @@
+### 2. MQTT Bridge Configuration
+
+The user can configure the MQTT Bridge using the `application.properties` file. This section describes the configuration 
+properties that can be used to configure the MQTT Bridge.
```
all on one line

ppatierno commented on README.md:
```diff
@@ -3,4 +3,135 @@
+
+The following table describes the configuration properties that can be used to configure the MQTT Bridge.
+
+| Setting                        | Description                                     | Deafult           |
```
```suggestion
| Setting                        | Description                                     | Default           |
```

ppatierno commented on README.md:
```diff
@@ -3,4 +3,135 @@
+| ------------------------------ | ----------------------------------------------- |-------------------|
+| bridge.id                      | ID of the bridge                                | my-bridge         |
+| mqtt.server.host               | Host address of the MQTT server                 | localhost         |
+| mqtt.server.port               | Port number of the MQTT server                  | 1883(default)     |
```
```suggestion
| mqtt.server.port               | Port number of the MQTT server                  | 1883              |
```

We already know it's a default

ppatierno commented on README.md:
```diff
@@ -3,4 +3,135 @@
+
+The TOMAR is a set of patterns the user provides defining how the MQTT Bridge maps MQTT topic names to Kafka topic names.
+A Mapping Rule is a model that contains an `MQTT topic pattern` and a `Kafka topic template`. 
+All the incoming MQTT message's topic should match an `MQTT topic pattern` in the `TOMAR` so that the bridge knows in which 
```
```suggestion
All the incoming MQTT message's topic should match an `MQTT topic pattern` in the `TOMAR` so that the bridge knows in which Kafka topic to produce this message. 
```

antonio-pedro99 commented:
I have applied some changes, please check how it is looking now.

ppatierno commented on README.md:
```diff
@@ -4,3 +4,112 @@
+
+The TOMAR is a set of patterns the user provides defining how the MQTT Bridge maps MQTT topic names to Kafka topic names.
+A Mapping Rule is a model that contains an `MQTT topic pattern` and a `Kafka topic template`. 
+All the incoming MQTT message's topic should match an `MQTT topic pattern` in the `TOMAR` so that the bridge knows in which Kafka topic to produce this message. This Kafka topic is defined by a template, `kafka Topic template`. 
```
```suggestion
All the incoming MQTT message's topic should match an `MQTT topic pattern` in the `TOMAR` so that the bridge knows in which Kafka topic to produce this message. 
This Kafka topic is defined by a template, `kafka Topic template`. 
```

ppatierno commented on README.md:
````diff
@@ -4,3 +4,112 @@
+  }
+]
+```
+The wildcard "#" represents one or more levels in the topic hierarchy. The wildcard "+" represents a single level in the topic hierarchy.
````
```suggestion
The wildcard "#" represents one or more levels in the topic hierarchy. 
The wildcard "+" represents a single level in the topic hierarchy.
```

Mentioning the placeholder is missing. Also explain that `#` can stay only at the end.
Finally a link to the MQTT spec explaining this would be useful to users.

ppatierno commented on README.md:
```diff
@@ -4,3 +4,112 @@
+
+### 2. MQTT Bridge Configuration
+
+The user can configure the MQTT Bridge using the `application.properties` file. This section describes the configuration properties that can be used to configure the MQTT Bridge.
```
```suggestion
The user can configure the MQTT Bridge using the `application.properties` file.
This section describes the configuration properties that can be used to configure the MQTT Bridge.
```

ppatierno commented on README.md:
```diff
@@ -4,3 +4,112 @@
+
+## MQTT bridge documentation
+
+This section provides information about the MQTT Bridge. It describes the MQTT Bridge, how it works, and how to configure it.
```
let's remove this section as well, but starting with the overview.

antonio-pedro99 commented:
As we have discussed via call, 
- I have removed the procedure to config the bridge. 
- I have removed the instructions to build, test and run the bridge for now.
- Added information about placeholders and format the README file in a more comprehensive way.

ppatierno commented on README.md:
```diff
@@ -4,3 +4,96 @@
+#### 1.1 How it works
+
+To enable a seamless integration between MQTT and Kafka, the MQTT Bridge provides a way to map MQTT topics to Kafka topics. 
+This mapping is done using a set of predefined patterns and kafka topic templates. The MQTT Bridge uses these patterns to map MQTT topics to Kafka topics. 
```
```suggestion
This mapping is done using a set of predefined patterns and Kafka topic templates.
The MQTT Bridge uses these patterns to map MQTT topics to Kafka topics. 
```

still one sentence, one line :-)

ppatierno commented on README.md:
```diff
@@ -4,3 +4,96 @@
+A Mapping Rule is a model that contains an `MQTT topic pattern` and a `Kafka topic template`. 
+All the incoming MQTT message's topic should match an `MQTT topic pattern` in the `TOMAR` so that the bridge knows in which Kafka topic to produce this message. This Kafka topic is defined by a template, `kafka Topic template`. However, if the incoming MQTT message's topic does not match any pattern in the `TOMAR`, the Bridge has a default Kafka topic where the incoming message will be mapped to, ...
+
+A valid TOMAR is a JSON file that contains an array of mapping rules. Each mapping rule is a JSON object that contains two properties: `mqttTopic` and `kafkaTopic`. The following is an example of a valid TOMAR:
```
```suggestion
A valid ToMaR is a JSON file that contains an array of mapping rules. 
Each mapping rule is a JSON object that contains two properties: `mqttTopic` and `kafkaTopic`.
The following is an example of a valid TOMAR:
```

One sentence, one line.

ppatierno commented on README.md:
````diff
@@ -4,3 +4,96 @@
+]
+```
+
+The wildcard "#" represents one or more levels in the MQTT topic hierarchy. The wildcard "+" represents a single level in the MQTT topic hierarchy.
````
```suggestion
The wildcard "#" represents one or more levels in the MQTT topic hierarchy.
The wildcard "+" represents a single level in the MQTT topic hierarchy.
```

One sentence, one line.

ppatierno commented on README.md:
```diff
@@ -4,3 +4,96 @@
+The wildcard "#" represents one or more levels in the MQTT topic hierarchy. The wildcard "+" represents a single level in the MQTT topic hierarchy.
+It worth's mentioning that it is the user's responsibility to adhere to the [MQTT 3.1.1 naming conventions](http://docs.oasis-open.org/mqtt/mqtt/v3.1.1/os/mqtt-v3.1.1-os.html#_Toc398718106) when defining the MQTT topic patterns.
+
+The parts with curly braces in the MQTT topic pattern are called `placeholders`. The MQTT Bridge will extract the values of these placeholders from the MQTT topic and use them to construct the Kafka topic name using the `kafkaTopic` template.
```
```suggestion
The parts with curly braces in the MQTT topic pattern are called `placeholders`.
The MQTT Bridge will extract the values of these placeholders from the MQTT topic and use them to construct the Kafka topic name using the `kafkaTopic` template.
```

One sentence, one line.

ppatierno commented on README.md:
```diff
@@ -4,3 +4,96 @@
+
+The parts with curly braces in the MQTT topic pattern are called `placeholders`. The MQTT Bridge will extract the values of these placeholders from the MQTT topic and use them to construct the Kafka topic name using the `kafkaTopic` template.
+
+Let's go through each rule to understand how the MQTT Bridge uses these rules to map MQTT topics to Kafka topics:
```
```suggestion
Let's go through each rule in the above example to understand how the MQTT Bridge uses these rules to map MQTT topics to Kafka topics:
```

ppatierno commented on README.md:
```diff
@@ -4,3 +4,96 @@
+
+Let's go through each rule to understand how the MQTT Bridge uses these rules to map MQTT topics to Kafka topics:
+
+1. MQTT Topic: "building/{building}/room/{room}/#" -> Kafka Topic: "building_{building}room{room}"
```
```suggestion
1. MQTT Topic: `building/{building}/room/{room}/#` -> Kafka Topic: `building_{building}_room_{room}`
```

ppatierno commented on README.md:
```diff
@@ -4,3 +4,96 @@
+
+1. MQTT Topic: "building/{building}/room/{room}/#" -> Kafka Topic: "building_{building}room{room}"
+
+    This rule maps MQTT topics of the form "building/{building}/room/{room}/#" to the Kafka topic "building_{building}room{room}". For example, if the MQTT topic is "building/A/room/1/floor/2", it will be mapped to the Kafka topic "building_A_room_1_floor_2".
```
```suggestion
    This rule maps MQTT topics of the form `building/{building}/room/{room}/#` to the Kafka topic `building_{building}_room_{room}`. 
    For example, if the MQTT topic is `building/A/room/1/floor/2`, it will be mapped to the Kafka topic `building_A_room_1_floor_2`.
```

ppatierno commented on README.md:
```diff
@@ -4,3 +4,96 @@
+
+    This rule maps MQTT topics of the form "building/{building}/room/{room}/#" to the Kafka topic "building_{building}room{room}". For example, if the MQTT topic is "building/A/room/1/floor/2", it will be mapped to the Kafka topic "building_A_room_1_floor_2".
+
+2. MQTT Topic: "sensors/#" -> Kafka Topic: "sensor_others"
```
```suggestion
2. MQTT Topic: `sensors/#` -> Kafka Topic: `sensor_others`
```

ppatierno commented on README.md:
```diff
@@ -4,3 +4,96 @@
+
+2. MQTT Topic: "sensors/#" -> Kafka Topic: "sensor_others"
+
+    This rule maps any MQTT topic starting with "sensors/" followed by any number of levels in the hierarchy to the Kafka topic "sensor_others". For example, if the MQTT topic is "sensors/temperature/living-room",  it will be mapped to the Kafka topic "sensor_others".
```
```suggestion
    This rule maps any MQTT topic starting with `sensors/` followed by any number of levels in the hierarchy to the Kafka topic `sensor_others`. 
    For example, if the MQTT topic is `sensors/temperature/living-room`,  it will be mapped to the Kafka topic `sensor_others`.
```

one sentence, one line

ppatierno commented on README.md:
```diff
@@ -4,3 +4,96 @@
+
+    This rule maps any MQTT topic starting with "sensors/" followed by any number of levels in the hierarchy to the Kafka topic "sensor_others". For example, if the MQTT topic is "sensors/temperature/living-room",  it will be mapped to the Kafka topic "sensor_others".
+
+3. MQTT Topic: "sensors/+/data" -> Kafka Topic: "sensor_data"
```
```suggestion
3. MQTT Topic: `sensors/+/data` -> Kafka Topic: `sensor_data`
```

ppatierno commented on README.md:
```diff
@@ -4,3 +4,96 @@
+
+3. MQTT Topic: "sensors/+/data" -> Kafka Topic: "sensor_data"
+
+    This rule maps MQTT topics of the form "sensors/+/data" to the Kafka topic "sensor_data". For example, if the MQTT topic is "sensors/temperature/data", it will be mapped to the Kafka topic "sensor_data".
```
```suggestion
    This rule maps MQTT topics of the form `sensors/+/data` to the Kafka topic `sensor_data`. 
    For example, if the MQTT topic is `sensors/temperature/data`, it will be mapped to the Kafka topic `sensor_data`.
```

one sentence, one line.

ppatierno commented on README.md:
```diff
@@ -4,3 +4,96 @@
+
+    This rule maps MQTT topics of the form "sensors/+/data" to the Kafka topic "sensor_data". For example, if the MQTT topic is "sensors/temperature/data", it will be mapped to the Kafka topic "sensor_data".
+
+These mapping rules demonstrate how different MQTT topics can be mapped to corresponding Kafka topics based on predefined patterns. The MQTT Bridge uses these rules to map MQTT topics to Kafka topics. 
```
```suggestion
```

I would just delete this line, I think we have said that already a few times.

ppatierno commented on README.md:
```diff
@@ -4,3 +4,96 @@
+    This rule maps MQTT topics of the form "sensors/+/data" to the Kafka topic "sensor_data". For example, if the MQTT topic is "sensors/temperature/data", it will be mapped to the Kafka topic "sensor_data".
+
+These mapping rules demonstrate how different MQTT topics can be mapped to corresponding Kafka topics based on predefined patterns. The MQTT Bridge uses these rules to map MQTT topics to Kafka topics. 
+The order in which the rules are defined is important. The MQTT Bridge will use the first rule that matches the MQTT topic. For example, if the MQTT topic is "sensors/temperature/data", it will be mapped to the Kafka topic "sensor_data" because the third rule matches the MQTT topic. If the third rule was not defined, the MQTT Bridge would use the second rule to map the MQTT topic to the Kafka top ...
```
but the application.properties file you are showing has "sensors/#" defined before "sensors/+/data" so it's the other way around as you are describing or?

ppatierno commented on README.md:
```diff
@@ -4,3 +4,96 @@
+
+### 2. MQTT Bridge Configuration
+
+The user can configure the MQTT Bridge using an `application.properties` file. This section describes the configuration properties that can be used to configure the MQTT Bridge. 
```
```suggestion
The user can configure the MQTT Bridge using an `application.properties` file.
This section describes the configuration properties that can be used to configure the MQTT Bridge. 
```

one sentence, one line

ppatierno commented on README.md:
```diff
@@ -4,3 +4,96 @@
+### 2. MQTT Bridge Configuration
+
+The user can configure the MQTT Bridge using an `application.properties` file. This section describes the configuration properties that can be used to configure the MQTT Bridge. 
+The MQTT bridge can be configured using the appropriate prefix. Example:
```
```suggestion
The MQTT bridge can be configured using the appropriate prefix.
Example:
```

ppatierno reviewed (changes requested):
Guess what ... a lot of fixes around "one sentence, one line" I mentioned during last call :-)

antonio-pedro99 commented:
I think it should be fine now.
Thanks for the review.

ppatierno commented on README.md:
```diff
@@ -4,3 +4,107 @@
+
+Let's go through each rule in the above example to understand how the MQTT Bridge uses these rules to map MQTT topics to Kafka topics:
+
+1. MQTT Topic: `"building/{building}/room/{room}/#"` -> Kafka Topic: `"building_{building}room{room}"`
```
```suggestion
1. MQTT Topic: `building/{building}/room/{room}/#` -> Kafka Topic: `building_{building}_room_{room}`
```

ppatierno commented on README.md:
```diff
@@ -4,3 +4,107 @@
+
+1. MQTT Topic: `"building/{building}/room/{room}/#"` -> Kafka Topic: `"building_{building}room{room}"`
+
+    This rule maps MQTT topics of the form `"building/{building}/room/{room}/#"` to the Kafka topic `"building_{building}room{room}"`. 
```
```suggestion
    This rule maps MQTT topics of the form `building/{building}/room/{room}/#` to the Kafka topic `building_{building}_room_{room}`. 
```

ppatierno commented on README.md:
```diff
@@ -4,3 +4,107 @@
+    This rule maps MQTT topics of the form `"building/{building}/room/{room}/#"` to the Kafka topic `"building_{building}room{room}"`. 
+    For example, if the MQTT topic is "building/A/room/1/floor/2", it will be mapped to the Kafka topic "building_A_room_1_floor_2".
+
+2. MQTT Topic: `"sensors/+/data"` -> Kafka Topic: `"sensor_data"`
```
```suggestion
2. MQTT Topic: `sensors/+/data` -> Kafka Topic: `sensor_data`
```

ppatierno commented on README.md:
```diff
@@ -4,3 +4,107 @@
+    This rule maps MQTT topics of the form `sensors/+/data` to the Kafka topic `sensor_data`. 
+    For example, if the MQTT topic is `sensors/temperature/data`, it will be mapped to the Kafka topic `sensor_data`.
+
+3. MQTT Topic: `"sensors/#"` -> Kafka Topic: `"sensor_others"`
```
```suggestion
3. MQTT Topic: `sensors/#` -> Kafka Topic: `sensor_others`
```

ppatierno commented on README.md:
```diff
@@ -4,3 +4,107 @@
+
+3. MQTT Topic: `"sensors/#"` -> Kafka Topic: `"sensor_others"`
+
+   This rule maps any MQTT topic starting with "sensors/" followed by any number of levels in the hierarchy to the Kafka topic "sensor_others".
```
```suggestion
   This rule maps any MQTT topic starting with `sensors/` followed by any number of levels in the hierarchy to the Kafka topic `sensor_others`.
```

ppatierno commented on README.md:
```diff
@@ -4,3 +4,107 @@
+3. MQTT Topic: `"sensors/#"` -> Kafka Topic: `"sensor_others"`
+
+   This rule maps any MQTT topic starting with "sensors/" followed by any number of levels in the hierarchy to the Kafka topic "sensor_others".
+   For example, if the MQTT topic is "sensors/temperature/living-room",  it will be mapped to the Kafka topic "sensor_others".
```
```suggestion
   For example, if the MQTT topic is `sensors/temperature/living-room`,  it will be mapped to the Kafka topic `sensor_others`.
```

antonio-pedro99 replied:
Fixed all of them in one commit.

ppatierno commented on README.md:
```diff
@@ -4,3 +4,107 @@
+   For example, if the MQTT topic is "sensors/temperature/living-room",  it will be mapped to the Kafka topic "sensor_others".
+
+
+The order in which the rules are defined is important. The MQTT Bridge will use the first rule that matches the MQTT topic. For example, if the MQTT topic is "sensors/temperature/data", it will be mapped to the Kafka topic "sensor_data" because `"sensors/+/data"` matches the MQTT topic before `"sensors/#"`. If exchange the positions of the rules, the MQTT Bridge would use the `"sensors/#"` to map ...
```
```suggestion
The order in which the rules are defined is important.
The MQTT Bridge will use the first rule that matches the MQTT topic.
For example, if the MQTT topic is `sensors/temperature/data`, it will be mapped to the Kafka topic `sensor_data` because `sensors/+/data` matches the MQTT topic before `sensors/#`. 
If we swap the positions of the rules, the MQTT Bridge would use the `sensors/#` to map the MQTT topic to the Kafka topic `sensor_others`.
```

ppatierno reviewed (approved):
LGTM. Thanks!

kyguy commented on README.md:
```diff
@@ -4,3 +4,110 @@
+The ToMaR is a set of patterns the user provides defining how the MQTT Bridge maps MQTT topic names to Kafka topic names.
+A Mapping Rule is a model that contains an `MQTT topic pattern` and a `Kafka topic template`. 
+All the incoming MQTT message's topic should match an `MQTT topic pattern` in the ToMaR so that the bridge knows in which Kafka topic to produce this message.
+This Kafka topic is defined by a template, `kafka Topic template`.
```
```suggestion
This Kafka topic is defined by a template, `Kafka Topic template`.
```

kyguy reviewed (approved):
Great work @antonio-pedro99
